### ISSUE TITLE
Incorrect `__aexit__` return-type-based control flow inference

### ISSUE THREAD

**Issue by llchan** [microsoft/pyright#9279]
**Describe the bug**
`__aexit__` with a `bool | None` return type seems to be handled incorrectly compared to `__aexit__` with a `bool` return type, despite being a superset of possible return types (i.e. both are capable of suppressing the exception).

**Code or Screenshots**
```python
import random


class Foo:
    async def __aenter__(self) -> None:
        return None

    async def __aexit__(self, *_: object) -> bool:
        return True


class Bar:
    async def __aenter__(self) -> None:
        return None

    async def __aexit__(self, *_: object) -> bool | None:
        return True


async def foo() -> int:
    async with Foo():
        if random.random() < 0.5:
            return 1
        else:
            raise RuntimeError()


async def bar() -> int:
    async with Bar():
        if random.random() < 0.5:
            return 1
        else:
            raise RuntimeError()
```
Both async context managers suppress exceptions, and both have `__aexit__` return type annotations that are capable of being truthy.  In the output below, we expect both `foo` and `bar` to be flagged for control flow that can reach the end of the function without returning an `int`, but only `foo` is flagged.

```
$ pyright demo.py
/path/demo.py
  /path/demo.py:20:20 - error: Function with declared return type "int" must return value on all code paths
    "None" is not assignable to "int" (reportReturnType)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
```console
$ pyright --version
pyright 1.1.385

```

**Comment by erictraut** [comment]
Pyright is working correctly here and in accordance with the Python typing spec. See [this section](https://typing.readthedocs.io/en/latest/spec/exceptions.html#context-managers) for details.

**Comment by llchan** [comment]
I see, in that case I agree `pyright` is working according to the spec.  I strongly disagree with the spec but will save that for a different forum...

In light of this spec, would it be reasonable to ask for `pyright` to flag a `__aexit__` that has the possibility of returning a truthy value when it is annotated as `bool | None`?  It seems that if we are treating that as a special case at the `async with` "call site" we probably should special case its returnable values as well.

**Comment by erictraut** [comment]
> I strongly disagree with the spec

Yeah, I probably wouldn't have chosen this either, but it was a convention established by mypy many years ago. It is now well established, so we're stuck with it now.

> would it be reasonable to ask for pyright to flag...

I don't think such a special case is merited. It would it create false positives in some cases because there are legitimate cases where context managers swallow exceptions in some contexts and not in others. And even if that weren't the case, I'm always reluctant to add special cases because they make a type checker more difficult to maintain and result in inconsistent and unpredictable behavior for users. There are rare cases where a special case is needed to accommodate a common pattern that is likely to result in a bug, but this isn't one of those cases.

**Comment by llchan** [comment]
The cases where a context manager can conditionally swallow exceptions should be annotated with a `bool` return type though.  What I'm suggesting is for it to basically say "hey you are potentially returning a truthy value where you have forgotten to mark `__aexit__` as `bool`".  Being silent seems dangerous because the program now has a possibility of exiting a function such as `bar` in the example above with a `None` value, despite everything type checking correctly.

I believe the only false positives would be `__aexit__`s that return `True` or `None`, which can pretty easily be changed to return `True`/`False`.  I'd rather those rare cases be modified than leave the possibility of incorrect type inference.
